### DISCUSSION TITLE
How to correctly use the Optional, and OptionalGetElement operators along with the If operator?

### DISCUSSION THREAD

**Discussion by Mulukulum** [onnx/onnx#6291] [Q&A]
## What I want to do:

I'm trying to create a simple onnx model using python to learn how to use the Optional types properly. 

My model takes either 2 or 3 inputs which are tensors of dimensions `[batch_size, 5]` and I want the model to output the sum of the given inputs. 

One of the reasons I want to do this is because I would eventually like to use a similar method to get an LLM Onnx model to have an optional KV cache input.

Let's say that we have upto 3 inputs `input_1`, `input_2` and the optional input, `input_3`

## What I've accomplished so far:
- Made my `input_3` to be an optional tensor. The type shows as `optional<float32[batch_size,5]>` in netron
- I put `input_3` into an OptionalHasElement operator.  I then feed this to an If Operator
- The `then_branch` of the If operator runs OptionalGetElement on `input_3`. This is then fed to an add node, which then adds everything together.
- My `else_branch` runs an OptionalGetElement operator on the output of the first addition **(which is NOT an optional type)** and passes that to the output.

## What I need help with:

1.) Is there a better way to do this? Is there any other way to implement optional inputs?

2.) I'm not really sure how the If Operator is supposed to work since the documentation says the If operator takes only one input. I tried passing it only one input but I got the error described in 3. I then tried feeding my `input_3` and the output of `input_1 + input_2` which is called `output_first_addition` to my If operator along with the inputs. In each subgraph, I have named the inputs the same thing. Netron screenshots attached for reference. I still get the error described in 3. What is the correct way to pass inputs to the If operator?

![image](https://github.com/user-attachments/assets/cd237f09-3129-4b15-88de-3e2132ec5914)
![image](https://github.com/user-attachments/assets/23babec1-7b2e-4812-92da-8c1d11355520)
![image](https://github.com/user-attachments/assets/97766600-6973-405e-94a8-7a39a897277a)

3.) My model check `onnx.checker.check_model` fails with the following error message.

```
Nodes in a graph must be topologically sorted, however input 'optional_get_element_node_output' of node: 
name: optional_add_node OpType: Add
 is not output of any previous nodes.

==> Context: Bad node spec for node. Name: if_node OpType: If
```
I cannot make sense of this message, since my `optional_get_element_output` is an output from my OptionalGetElement within my `then_branch` of the If operator. I suspect this is because of something I am doing wrong with the If operator but I could not find any other example implementations. 

Are there any example implementations I can refer to, specifically involving the If and the Optional operators? I'd appreciate any help on this issue. Thanks!

Code for reference:
```python
import onnx
model_name = 'model.onnx'
input_1 = onnx.helper.make_tensor_value_info('input_1', onnx.TensorProto.FLOAT, ["batch_size", 5])
input_2 = onnx.helper.make_tensor_value_info('input_2', onnx.TensorProto.FLOAT, ["batch_size", 5])
output_first_addition = onnx.helper.make_tensor_value_info('output_first_addition', onnx.TensorProto.FLOAT, ["batch_size", 5])


optional_input_type_proto = onnx.helper.make_tensor_type_proto(onnx.TensorProto.FLOAT, ["batch_size", 5])
input_3 = onnx.helper.make_value_info(name='input_3', type_proto=onnx.TypeProto(optional_type=onnx.TypeProto.Optional(elem_type=optional_input_type_proto)))

subgraph_output = onnx.helper.make_tensor_value_info('optional_add_node_output', onnx.TensorProto.FLOAT, ["batch_size", 5])
optional_add_node_output = onnx.helper.make_tensor_value_info('optional_add_node_output', onnx.TensorProto.FLOAT, ["batch_size", 5])

add_node = onnx.helper.make_node('Add', ['input_1', 'input_2'], ['output_first_addition'], name='add_node')
final_output = onnx.helper.make_tensor_value_info('output', onnx.TensorProto.FLOAT, ["batch_size", 5])

optional_has_element_node = onnx.helper.make_node('OptionalHasElement', ['input_3'], ['optional_has_element_node_output'], name='optional_has_element_node')
optional_get_element_node = onnx.helper.make_node('OptionalGetElement', ['input_3'], ['optional_get_element_node_output'], name='optional_get_element_node')
optional_add_node = onnx.helper.make_node('Add', ['output_first_addition','optional_get_element_node_output'], ['optional_add_node_output'], name='optional_add_node')

optional_get_element_else_branch = onnx.helper.make_node('OptionalGetElement',['output_first_addition'], ['optional_add_node_output'], name='optional_get_element_else_branch')

#if_node : onnx.NodeProto = onnx.helper.make_node('If', ['optional_has_element_node_output','output_first_addition','input_3'], ['output'], name='if_node')

if_node : onnx.NodeProto = onnx.helper.make_node('If', ['optional_has_element_node_output',], ['output'], name='if_node')


then_branch = if_node.attribute.add(g = onnx.helper.make_graph([optional_add_node, optional_get_element_node], 'then_branch', [output_first_addition, input_3], [optional_add_node_output]),
                                    name='then_branch',
                                    type=onnx.AttributeProto.GRAPH)

else_branch = if_node.attribute.add(g=onnx.helper.make_graph(nodes=[optional_get_element_else_branch], name='else_branch', inputs=[output_first_addition], outputs=[optional_add_node_output]),
                                    name='else_branch',
                                    type=onnx.AttributeProto.GRAPH)
onnx.ValueInfoProto
execution_graph = onnx.helper.make_graph(
    [add_node, optional_has_element_node, if_node],
    'addition',
    [input_1, input_2, input_3],
    [final_output]
)

model = onnx.helper.make_model(execution_graph)
#onnx.checker.check_model(model)
onnx.save(model, model_name)


```

**Comment by Mulukulum** [comment]
okay, well seems like the error goes away when I switched the positions of `optional_add_node` and `optional_get_element_node` in this snippet.

```
then_branch = if_node.attribute.add(g = onnx.helper.make_graph([optional_add_node, optional_get_element_node], 'then_branch', [output_first_addition, input_3], [optional_add_node_output]),
                                    name='then_branch',
                                    type=onnx.AttributeProto.GRAPH)
```

When trying to load my model using onnxruntime I get the following error
```
1 : FAIL : Load model from model.onnx failed:Node (if_node) Op (If) [TypeInferenceError] Graph attribute inferencing failed: Size mismatch validating subgraph inputs. Got 0 inputs but subgraph has 2 inputs and requires 2 inputs. Either provide all subgraph inputs, or just the required inputs.
```

**Comment by Mulukulum** [comment]
okay, seems like according to #3255 the issue is that if subgraphs shouldn't have inputs specified. I have removed that and I'm now able to run inference.

I would still appreciate if anyone from the community has example implementations involving If, Optionals or if anyone has ideas for how this can be achieved better. Thanks!

**Comment by justinchuby** [comment]
cc @gramalingam an interesting example using optional types
